Implement Observable#reduce

maraisr opened the pull request:
#### 5799af7ea9f2c0f39b9857651240967129be2900
<pre>
Implement Observable#reduce
<a href="https://bugs.webkit.org/show_bug.cgi?id=284287">https://bugs.webkit.org/show_bug.cgi?id=284287</a>

Reviewed by Darin Adler.

This change introduces the `.reduce` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.

Unlike our MapperCallback, we need to pass an accumulator as well. As
such, we also introduce a new ReducerCallback IDL.

Spec: &lt;<a href="https://wicg.github.io/observable/#dom-observable-reduce">https://wicg.github.io/observable/#dom-observable-reduce</a>&gt;

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any.js: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any.worker-expected.txt: Rebaseline.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverReduce.cpp: Added.
(WebCore::createInternalObserverOperatorReduce):
* Source/WebCore/dom/InternalObserverReduce.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::reduce):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes `Promise&lt;any&gt; reduce(reducer, initialValue, options)` as a promise-returning operator.
* Source/WebCore/dom/ReducerCallback.h: Added.
* Source/WebCore/dom/ReducerCallback.idl: Added.

Canonical link: <a href="https://commits.webkit.org/287595@main">https://commits.webkit.org/287595@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc06f250f53bf7c3067cb825c79ed54c0cca64c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59202 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31176 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62692 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20504 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73038 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43000 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27192 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29637 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71219 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86149 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5249 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70963 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7595 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70207 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17481 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14199 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13150 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12901 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10743 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9028 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any.js:
Upstreamed test changes here: https://github.com/web-platform-tests/wpt/pull/49588 and https://github.com/web-platform-tests/wpt/pull/49589 as such considering this change a rebaseline.

maraisr commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -169,6 +171,15 @@ void Observable::some(ScriptExecutionContext& context, Ref<PredicateCallback>&&
+    // IDL for 'optional any' defaults to 'jsUndefined',
+    // so std::optional is always present. If undefined, set to nullopt.
+    if (initialValue->isUndefined())
+        initialValue = std::nullopt;
```
This felt a little funky; the only way I got an `IDLOptional` to generate in the JS* classes, was with `optional object` than `optional any`.

Does `optional any` suggest that; `reduce(fn)` and `reduce(fn, undefined)` are equivalent, and if so, is checking for "undefined" enough to suffice the spec requirement in [point 7](https://wicg.github.io/observable/#dom-observable-reduce)

darinadler replied:
The question of whether `reduce(fn)` and `reduce(fn, undefined)` are equivalent is best answered with test cases so we know it’s interoperable. For as long as I can remember I’ve seen questions about whether `undefined` and a missing argument behave the same, and the answer has been different in different contexts. For example, things built into the JavaScript language seem to definitely treat them the same, but not sure DOM always does. If we have a test case, then we can worry about how we are coding it and not whether the behavior is correct.

In terms of how we write the code, if the bindings generated code never passes `std::nullopt`, then we should just use `JSC::JSValue`, and not involve `std::optional`. We can always switch to using `std::optional` later if we change how the bindings generate. It’s not helpful to use rvalue references with `std::optional<JSC::JSValue>`, by the way, since it’s all scalars and no helpful meaning to the move semantic, so we should probably not use `&&`.

darinadler commented on Source/WebCore/dom/InternalObserverReduce.cpp:
```diff
@@ -0,0 +1,145 @@
+    {
+    }
+
+    uint64_t m_idx { 0 };
```
WebKit coding style calls this `m_index`, not `m_idx`, because we use words instead of abbreviations.

darinadler commented on Source/WebCore/dom/InternalObserverReduce.h:
```diff
@@ -0,0 +1,40 @@
+class DeferredPromise;
+class Observable;
+class ScriptExecutionContext;
+class ReducerCallback;
```
Sort these alphabetically, please: `ReducerCallback` before `ScriptExecutionContext`.

darinadler commented on Source/WebCore/dom/Observable.h:
```diff
@@ -72,6 +73,7 @@ class Observable final : public ScriptWrappable, public RefCounted<Observable> {
     void find(ScriptExecutionContext&, Ref<PredicateCallback>&&, const SubscribeOptions&, Ref<DeferredPromise>&&);
     void every(ScriptExecutionContext&, Ref<PredicateCallback>&&, const SubscribeOptions&, Ref<DeferredPromise>&&);
     void some(ScriptExecutionContext&, Ref<PredicateCallback>&&, const SubscribeOptions&, Ref<DeferredPromise>&&);
+    void reduce(ScriptExecutionContext&, Ref<ReducerCallback>&&, std::optional<JSC::JSValue>&&, const SubscribeOptions&, Ref<DeferredPromise>&&);
```
As I said above, I think this can just be `const std::optional<JSC::JSValue>&` or `std::optional<JSC::JSValue>` or `JSC::JSValue`, nothing that needs to be moved or reason to use `&&`.

darinadler commented on Source/WebCore/dom/InternalObserverReduce.cpp:
```diff
@@ -0,0 +1,145 @@
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
```
Need to indent this one space so the `*` line up.

maraisr replied:
Cheers for catching this; I have also raised https://github.com/WebKit/WebKit/pull/37666  to cleanup the other ones I've copied and pasted. Did a `rg "Marais"` and checked all the _my_ ones.

darinadler commented on Source/WebCore/dom/InternalObserverReduce.h:
```diff
@@ -0,0 +1,40 @@
+class ReducerCallback;
+struct SubscribeOptions;
+
+void createInternalObserverOperatorReduce(ScriptExecutionContext&, Observable&, Ref<ReducerCallback>&&, std::optional<JSC::JSValue>&&, const SubscribeOptions&, Ref<DeferredPromise>&&);
```
As I said above, I think this can just be `const std::optional<JSC::JSValue>&` or `std::optional<JSC::JSValue>` or `JSC::JSValue`, nothing that needs to be moved or reason to use `&&`.

maraisr commented on Source/WebCore/dom/InternalObserverReduce.cpp:
```diff
@@ -0,0 +1,145 @@
+    uint64_t m_idx { 0 };
+    Ref<AbortSignal> m_signal;
+    Ref<ReducerCallback> m_callback;
+    std::optional<JSC::JSValue> m_accumulator;
```
@darinadler @cdumez Should this really be

```suggestion
    std::optional<JSValueInWrappedObject> m_accumulator;
```

over slack you mentioned that with like with [last](https://github.com/WebKit/WebKit/blob/a542f6c9c8b1f812fe86dea9055334e1ddf4a0d1/Source/WebCore/dom/InternalObserverLast.cpp#L86) nothing is keeping the value alive?

darinadler replied:
Oh, yes this is wrong, can’t just put a `JSValue` in an object without wrapping it. It’s not enough to just use `JSValueInWrappedObject`.

maraisr replied:
What would you suggest as per a fix?

We have the visitAdditionalChildren wirings already, and I see `JSValueInWrappedObject` has a `visit` method, can I use that and visit in our existing visitAdditionalChildren?

darinadler replied:
Yes, I think so …

maraisr replied:
There is a comment that kind of makes me want to use `set`, or rather when would it be "cleared from under you" as per the setWeakly (default) behaviour?

https://github.com/WebKit/WebKit/blob/c112fe9adf5d1c10ca6abf2542900dda02286f03/Source/WebCore/bindings/js/JSValueInWrappedObject.h#L49-L51

If I should be using `set`, where would I get a `JSC::JSCell`. Most of the sample code I can find around, simply use the `setWeakly` (default) behaviour, so maybe it is okay.

maraisr replied:
Perhaps the JSCell (owner argument), can be the `Observable` instance object?

maraisr replied:
Oh wait never mind, we are calling `visit`, in our visitAdditionalChildren — so it wont be cleaned from under us, and its safe. Pushing this in a moment.

maraisr replied:
Raised PR fix for the `last` operator: https://github.com/WebKit/WebKit/pull/37670

maraisr commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -169,6 +171,16 @@ void Observable::some(ScriptExecutionContext& context, Ref<PredicateCallback>&&
+{
+    // IDL for 'optional any' defaults to 'undefined'.
+    // We need to use std::optional to distinguish between 'undefined' and 'not provided'.
+    std::optional<JSC::JSValue> internalInitialValue;
```
This is mainly so the reducer does not call for the first value if the initial value is undefined/not passed, as per spec.

If we did not use `std::optional`, we'd have more machinery to detect a reducer that returns undefined, and not the first value. Maybe the m_index can be abused for this? Because we set the accumulator from the reducer's result, and check that again. And Id rather not store 2 data members, one for the intial value, and one for the accumulator. 

So think—and could be wrong—that the semantics of our operator constructor, should have a concept of set and unset initial values; thus using std::optional.

darinadler replied:
To prove this we need to add test cases that stress test these edge cases and fine distinctions. Reasoning about the source code won’t be enough to get this right.

darinadler commented on Source/WebCore/dom/InternalObserverReduce.h:
```diff
@@ -0,0 +1,40 @@
+class ScriptExecutionContext;
+struct SubscribeOptions;
+
+void createInternalObserverOperatorReduce(ScriptExecutionContext&, Observable&, Ref<ReducerCallback>&&, const std::optional<JSC::JSValue>&, const SubscribeOptions&, Ref<DeferredPromise>&&);
```
I think this can take `JSC::JSValue` instead of `const std::optional<JSC::JSValue>&`. Not sure that the optional is really working for us here. Could just use `undefined` at first maybe? I guess we can leave this as-is for now while we create test cases that try to tickle `undefined` vs. no value for initial value.

maraisr replied:
I think this best explains it: https://github.com/WebKit/WebKit/pull/37630#discussion_r1876908305

We dont want to call the reducer if the initial value is undefined/not-passed. There is a certain convenience imposed by std::optional.

But I also hear you, perhaps its better we localize this logic closer to where it matters. Ill move it.

maraisr replied:
Okay pushed a change that moves this down to where it matters. Thanks for the nudge!

maraisr committed:
Implement Observable#reduce
https://bugs.webkit.org/show_bug.cgi?id=284287

Reviewed by Darin Adler.

This change introduces the `.reduce` promise-returning operator
for Observables, using the InternalObserver and subscribing immediately.

Unlike our MapperCallback, we need to pass an accumulator as well. As
such, we also introduce a new ReducerCallback IDL.

Spec: <https://wicg.github.io/observable/#dom-observable-reduce>

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any.js: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-reduce.any.worker-expected.txt: Rebaseline.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverReduce.cpp: Added.
(WebCore::createInternalObserverOperatorReduce):
* Source/WebCore/dom/InternalObserverReduce.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::reduce):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes `Promise<any> reduce(reducer, initialValue, options)` as a promise-returning operator.
* Source/WebCore/dom/ReducerCallback.h: Added.
* Source/WebCore/dom/ReducerCallback.idl: Added.

Canonical link: https://commits.webkit.org/287595@main
